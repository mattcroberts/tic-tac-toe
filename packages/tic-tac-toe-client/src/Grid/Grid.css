.Grid {
  display: flex;
  flex-direction: column;
  height: 100vh;
  text-align: center;
}

.Grid-row {
  display: flex;
  height: 100%;
  flex-direction: row;
}
