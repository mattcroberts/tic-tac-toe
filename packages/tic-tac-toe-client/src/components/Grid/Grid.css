.root {
    display: flex;
    flex-direction: column;
    flex: 1;
    text-align: center;
    position: relative;
    user-select: none;
}

.row {
    display: flex;
    flex: 1;
}
