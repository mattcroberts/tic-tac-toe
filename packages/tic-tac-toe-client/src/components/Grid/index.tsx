import * as React from "react";
import {
    GridItem as IGridItem,
    Player as IPlayer
} from "../../../typings/types";
import GridItem from "../GridItem";
import Winner from "../Winner";

import style from "./Grid.css";

export interface IProps {
    id: string;
    grid: IGridItem[][];
    currentPlayer: IPlayer;
    controllingPlayer: IPlayer;
    winner?: IPlayer | null;
    isDraw: boolean;
    size: number;
    executeTurn: ({ variables: ExecuteTurnMutationArgs }) => any;
}

class Grid extends React.Component<IProps> {
    public static defaultProps = {
        isDraw: false,
        winner: null
    };

    constructor(props: IProps) {
        super(props);
        this.onItemClick = this.onItemClick.bind(this);
    }

    public render() {
        return (
            <div className={style.root}>
                {this.renderGrid()}
                {this.props.winner || this.props.isDraw ? (
                    <Winner winner={this.props.winner} />
                ) : null}
            </div>
        );
    }

    public onItemClick(x: number, y: number) {
        const { controllingPlayer, currentPlayer, executeTurn } = this.props;

        if (
            this.props.grid[x][y].player === null &&
            controllingPlayer !== null &&
            controllingPlayer.symbol === this.props.currentPlayer.symbol
        ) {
            executeTurn({
                variables: {
                    id: this.props.id,
                    playerId: currentPlayer.id,
                    x,
                    y
                }
            });
        }
    }

    private renderGrid() {
        return this.props.grid.map((row, rowN) => (
            <div className={style.row} key={`${rowN.toString()}`}>
                {row.map((item, colN) => (
                    <GridItem
                        key={`${rowN}-${colN}`}
                        colN={rowN}
                        rowN={colN}
                        onItemClick={this.onItemClick}
                        itemState={item}
                        isFinished={
                            this.props.winner !== null || this.props.isDraw
                        }
                        position={{
                            bottom: rowN === 2,
                            left: colN === 0,
                            right: colN === 2,
                            top: rowN === 0
                        }}
                    />
                ))}
            </div>
        ));
    }
}

export default Grid;
