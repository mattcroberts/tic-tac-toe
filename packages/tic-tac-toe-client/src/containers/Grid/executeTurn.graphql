mutation executeTurn($id: ID!, $playerId: ID!, $x: Int!, $y: Int!) {
    executeTurn(id: $id, playerId: $playerId, x: $x, y: $y) {
        id
        gridItems {
            id
            player {
                id
                symbol
            }
        }
        currentPlayer {
            id
            symbol
        }
        winner {
            id
            symbol
        }
        isFinished
    }
}
