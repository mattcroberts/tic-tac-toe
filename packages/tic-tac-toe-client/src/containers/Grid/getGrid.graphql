query GET_GRID($id: ID!) {
    grid(id: $id) {
        id
        gridItems {
            player {
                id
                symbol
            }
        }
        currentPlayer {
            id
            symbol
        }
        winner {
            id
            symbol
        }
        isFinished
        size
        gameUrls {
            NAUGHT
            CROSS
        }
        players {
            id
            symbol
        }
    }
}
