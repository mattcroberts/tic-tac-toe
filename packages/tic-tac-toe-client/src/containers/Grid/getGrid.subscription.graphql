subscription GET_GRID($id: ID!) {
    gridUpdated(id: $id) {
        id
        gridItems {
            player
        }
        currentPlayer {
            id
            symbol
        }
        winner {
            id
            symbol
        }
        isFinished
        size
        gameUrls {
            NAUGHT
            CROSS
        }
        players {
            id
            symbol
        }
    }
}
