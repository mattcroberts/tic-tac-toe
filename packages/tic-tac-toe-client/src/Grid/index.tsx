// import { merge } from "lodash";
import * as React from "react";

import GridItem, { IItemState } from "../GridItem";

import "./Grid.css";

export enum Player {
  CROSS = "X",
  NAUGHT = "0"
}
export interface IGridProps {
  grid: IItemState[][];
  currentPlayer?: Player;
  hasWon?: boolean;
}

interface IGridState {
  grid: IItemState[][];
  currentPlayer: Player;
  hasWon: boolean;
  winner: Player | null;
}

class Grid extends React.Component<IGridProps, IGridState> {
  public static defaultProps = {
    currentPlayer: Player.CROSS,
    hasWon: false
  };
  constructor(props: IGridProps) {
    super(props);
    this.state = {
      currentPlayer: props.currentPlayer as Player,
      grid: props.grid,
      hasWon: props.hasWon as boolean,
      winner: null
    };

    this.hasWon = this.hasWon.bind(this);
    this.onItemClick = this.onItemClick.bind(this);
  }

  public render() {
    return (
      <div className="Grid">
        {this.state.hasWon ? `${this.state.winner} HAS WON` : this.renderGrid()}
      </div>
    );
  }

  private hasWon(player: Player, grid: IItemState[][]): boolean {
    const rowWin = grid.reduce((hasWon, row, i, all) => {
      return (
        hasWon ||
        row.every(({ player: p }, j) => {
          return p === player;
        })
      );
    }, false);

    let colWin = false;
    const forwardDiag: IItemState[] = [];
    const backwardDiag: IItemState[] = [];
    for (let i = 0; i < grid.length; i++) {
      const col: IItemState[] = [];
      forwardDiag.push(grid[i][i]);
      backwardDiag.push(grid[i][grid.length - i - 1]);

      for (let j = 0; j < grid[i].length; j++) {
        col.push(grid[j][i]);
      }

      colWin =
        col[0].player === player &&
        col.every(({ player: p }) => {
          return p === col[0].player;
        });

      if (colWin) {
        break;
      }
    }

    const forwardDiagWin =
      forwardDiag[0].player === player &&
      forwardDiag.every(({ player: p }) => {
        return p === forwardDiag[0].player;
      });

    const backwardDiagWin =
      backwardDiag[0].player === player &&
      backwardDiag.every(({ player: p }) => {
        return p === backwardDiag[0].player;
      });

    return rowWin || colWin || forwardDiagWin || backwardDiagWin;
  }

  private onItemClick(colN: number, rowN: number) {
    if (!this.state.hasWon) {
      this.setState(prevState => {
        const nextGrid = Array.from(prevState.grid);
        nextGrid[colN][rowN] = { player: prevState.currentPlayer };
        const crossHasWon = this.hasWon(Player.CROSS, nextGrid);
        const naughtHasWon = this.hasWon(Player.NAUGHT, nextGrid);

        const newState = {
          ...prevState,
          grid: nextGrid,
          hasWon: crossHasWon || naughtHasWon,
          winner: crossHasWon
            ? Player.CROSS
            : naughtHasWon
              ? Player.NAUGHT
              : null
        };

        if (!crossHasWon && !naughtHasWon) {
          newState.currentPlayer =
            prevState.currentPlayer === Player.CROSS
              ? Player.NAUGHT
              : Player.CROSS;
        }

        return newState;
      });
    }
  }

  private renderGrid() {
    return this.state.grid.map((row, rowN) => (
      <div className="Grid-row" key={`${rowN}`}>
        {row.map((item, colN) => (
          <React.Fragment>
            <span>
              {rowN},
              {colN}
            </span>
            <GridItem
              colN={rowN}
              rowN={colN}
              key={`${colN}-${rowN}`}
              onItemClick={this.onItemClick}
              itemState={item}
            />
          </React.Fragment>
        ))}
      </div>
    ));
  }
}

export default Grid;
