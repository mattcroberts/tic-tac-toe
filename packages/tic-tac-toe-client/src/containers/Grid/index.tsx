import * as React from "react";
import { graphql, Subscription } from "react-apollo";
import { withRouter, RouteComponentProps } from "react-router-dom";
import { compose } from "recompose";
import { Grid as IGrid } from "../../../typings/types";
import GridPage from "../../pages/Grid";

import * as EXECUTE_TURN from "./executeTurn.graphql";
import * as GET_GRID from "./getGrid.graphql";
import * as GET_GRID_SUBS from "./getGrid.subscription.graphql";
import ErrorPage from "../../pages/Error";

export interface IProps extends RouteComponentProps<{ playerId: string }> {
    executeTurn: (
        options: {
            variables: {
                id: string;
                playerId: string;
                x: number;
                y: number;
            };
        }
    ) => IGrid;
    isMultiplayer: boolean;
    invitedPlayerId: string;
    data: {
        error: any;
        grid: IGrid;
        loading: boolean;
    };
}

export class GridContainer extends React.Component<IProps> {
    public state = {};

    constructor(props: IProps) {
        super(props);
        this.onItemClick = this.onItemClick.bind(this);
    }

    public render() {
        const {
            invitedPlayerId,
            data: { error, grid, loading }
        } = this.props;

        if (loading) {
            return <p>Loading...</p>;
        }

        if (!grid) {
            return <ErrorPage message="Game not found" />;
        }

        if (!grid.players || grid.players.length < 2) {
            return <p>Player error</p>;
        }

        const controllingPlayer = this.props.isMultiplayer
            ? this.getControllingPlayer(grid, invitedPlayerId)
            : grid.currentPlayer;

        if (!controllingPlayer) {
            return <p>Player not found</p>;
        }

        if (error) {
            return <p>{error.message}</p>;
        }

        return (
            <Subscription
                subscription={GET_GRID_SUBS}
                variables={{ id: grid.id }}
            >
                {({
                    data: { gridUpdated } = { gridUpdated: undefined },
                    loading: subsLoading
                }: {
                    data: { gridUpdated: IGrid | undefined };
                    loading: boolean;
                }) => {
                    const g =
                        subsLoading || gridUpdated === undefined
                            ? grid
                            : gridUpdated;
                    return (
                        <GridPage
                            grid={g.gridItems}
                            currentPlayer={g.currentPlayer}
                            winner={g.winner}
                            isDraw={g.winner === null && g.isFinished}
                            size={g.size}
                            onItemClick={this.onItemClick}
                            gameUrls={g.gameUrls}
                            isMultiplayer={this.props.isMultiplayer}
                            controllingPlayer={controllingPlayer}
                        />
                    );
                }}
            </Subscription>
        );
    }

    private getControllingPlayer(grid: IGrid, invitedPlayerId: string) {
        if (!grid.players || grid.players.length < 2) {
            return null;
        }

        return (
            grid.players.find(p => p !== null && p.id === invitedPlayerId) ||
            grid.players[0]
        );
    }

    private onItemClick(x: number, y: number): void {
        const {
            executeTurn,
            data: { grid }
        } = this.props;

        const controllingPlayer = this.props.isMultiplayer
            ? this.getControllingPlayer(grid, this.props.invitedPlayerId)
            : this.props.data.grid.currentPlayer;

        if (
            grid.gridItems[x][y].player === null &&
            controllingPlayer !== null &&
            controllingPlayer.symbol === grid.currentPlayer.symbol
        ) {
            executeTurn({
                variables: {
                    id: grid.id,
                    playerId: controllingPlayer.id,
                    x,
                    y
                }
            });
        }
    }
}

const enhance = compose<{}, IProps>(
    withRouter,
    graphql(GET_GRID, {
        options: (props: any) => ({
            variables: {
                id: props.match.params.gameId
            }
        })
    }),
    graphql(EXECUTE_TURN, {
        name: "executeTurn"
    })
);

export default enhance(GridContainer);
