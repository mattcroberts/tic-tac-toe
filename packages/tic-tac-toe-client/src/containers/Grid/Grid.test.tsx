import * as React from "react";
import { mount } from "enzyme";
import { merge } from "lodash";
import { GridContainer } from ".";

import { createGrid } from "../../../test/utils";
import Grid from "../../pages/Grid";
import { Symbol as ISymbol } from "../../../typings/types";
import { MockedProvider } from "react-apollo/test-utils";
import * as GET_GRID_SUBS from "./getGrid.subscription.graphql";
import { MemoryRouter } from "react-router";

const mocks = [
    {
        request: {
            query: GET_GRID_SUBS
        },
        result: {
            data: {
                gridUpdated: {}
            }
        }
    }
];
describe("GridContainer", () => {
    const defaultProps = {
        history: undefined as any,
        location: undefined as any,
        match: {
            params: { gameId: { gameId: "123", playerId: "456" } }
        } as any,
        isMultiplayer: false,
        invitedPlayerId: "123",
        data: {
            error: undefined,
            grid: {
                gameUrls: {},
                currentPlayer: {
                    id: "abc",
                    symbol: ISymbol.NAUGHT
                },
                gridItems: createGrid(3),
                id: "abc123",
                isFinished: false,
                size: 9,
                players: [
                    { id: "abc", symbol: ISymbol.NAUGHT },
                    { id: "123", symbol: ISymbol.CROSS }
                ]
            },
            loading: false
        },
        executeTurn: jest.fn()
    };

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it("renders without crashing", () => {
        const grid = mount(
            <MockedProvider mocks={mocks} addTypename={false}>
                <MemoryRouter>
                    <GridContainer {...defaultProps} />
                </MemoryRouter>
            </MockedProvider>
        ).find(Grid);

        expect(grid).toMatchSnapshot();
    });

    it.skip("renders loading state", () => {
        const props = {
            ...defaultProps,
            data: {
                ...defaultProps.data,
                loading: true
            }
        };

        const grid = mount(
            <MockedProvider mocks={[]} addTypename={false}>
                <MemoryRouter>
                    <GridContainer {...props} />
                </MemoryRouter>
            </MockedProvider>
        ).find(Grid);
        expect(grid).toMatchSnapshot();
    });
});
