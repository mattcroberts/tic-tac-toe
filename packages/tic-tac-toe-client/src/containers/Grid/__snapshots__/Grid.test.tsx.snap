// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridContainer renders loading state 1`] = `
<p>
  Loading...
</p>
`;

exports[`GridContainer renders without crashing 1`] = `
<GridPage
  currentPlayer="NAUGHT"
  grid={
    Array [
      Array [
        Object {
          "id": "0_0",
          "player": null,
        },
        Object {
          "id": "0_1",
          "player": null,
        },
        Object {
          "id": "0_2",
          "player": null,
        },
      ],
      Array [
        Object {
          "id": "1_0",
          "player": null,
        },
        Object {
          "id": "1_1",
          "player": null,
        },
        Object {
          "id": "1_2",
          "player": null,
        },
      ],
      Array [
        Object {
          "id": "2_0",
          "player": null,
        },
        Object {
          "id": "2_1",
          "player": null,
        },
        Object {
          "id": "2_2",
          "player": null,
        },
      ],
    ]
  }
  isDraw={false}
  onItemClick={[Function]}
  size={9}
/>
`;
